x-pack/solutions/observability/plugins/observability/server/services/related_dashboards_client.ts

```diff
@@ -0,0 +1,267 @@
+import type { AlertData } from './alert_data';
+
+type Dashboard = SavedObjectsFindResult<DashboardAttributes>;
+export class RelatedDashboardsClient {
+  private dashboardsById = new Map<string, Dashboard>();
+  private alert: AlertData | null = null;
+
+  constructor(
+    private logger: Logger,
+    private dashboardClient: IContentClient<Dashboard>,
+    private alertsClient: InvestigateAlertsClient,
+    private alertId: string
+  ) {}
+
+  async fetchSuggestedDashboards(): Promise<{ suggestedDashboards: RelatedDashboard[] }> {
+    const allRelatedDashboards = new Set<RelatedDashboard>();
+    const relevantDashboardsById = new Map<string, RelatedDashboard>();
+    const [alert] = await Promise.all([
+      this.alertsClient.getAlertById(this.alertId),
+      this.fetchAllDashboards(),
+    ]);
+    this.alert = alert;
+    if (!this.alert) {
+      return { suggestedDashboards: [] };
+    }
+    const index = await this.getRuleQueryIndex();
+    const relevantRuleFields = this.alert.getRelevantRuleFields();
+    const relevantAlertFields = this.alert.getRelevantAADFields();
+    const allRelevantFields = new Set([...relevantRuleFields, ...relevantAlertFields]);
+
+    if (index) {
+      const { dashboards } = this.getDashboardsByIndex(index);
+      dashboards.forEach((dashboard) => allRelatedDashboards.add(dashboard));
+    }
+    if (allRelevantFields.size > 0) {
+      const { dashboards } = this.getDashboardsByField(Array.from(allRelevantFields));
+      dashboards.forEach((dashboard) => allRelatedDashboards.add(dashboard));
+    }
+    allRelatedDashboards.forEach((dashboard) => {
+      const dedupedPanels = this.dedupePanels([
+        ...(relevantDashboardsById.get(dashboard.id)?.relevantPanels || []),
+        ...dashboard.relevantPanels,
+      ]);
+      const relevantPanelCount = dedupedPanels.length;
+      relevantDashboardsById.set(dashboard.id, {
+        ...dashboard,
+        matchedBy: {
+          ...relevantDashboardsById.get(dashboard.id)?.matchedBy,
+          ...dashboard.matchedBy,
+        },
+        relevantPanelCount,
+        relevantPanels: dedupedPanels,
+      });
+    });
+    return { suggestedDashboards: Array.from(relevantDashboardsById.values()) };
+  }
+
+  async fetchDashboards(page: number) {
+    const perPage = 1000;
+    const dashboards = await this.dashboardClient.search(
+      { limit: perPage, cursor: `${page}` },
+      { spaces: ['*'] }
+    );
+    const {
+      result: { hits },
+    } = dashboards;
+    hits.forEach((dashboard: Dashboard) => {
+      this.dashboardsById.set(dashboard.id, dashboard);
+    });
+    const fetchedUntil = (page - 1) * perPage + dashboards.result.hits.length;
+
+    if (dashboards.result.pagination.total <= fetchedUntil) {
+      return;
+    }
+    await this.fetchDashboards(page + 1);
+  }
+
+  async fetchAllDashboards() {
+    await this.fetchDashboards(1);
+  }
+
+  getDashboardsByIndex(index: string): {
```

Comment:
Maybe we should expose the [`SavedObjectsFindOptions.hasReferences` property](https://github.com/elastic/kibana/blob/59e606cdc49e6500bfd7a9a6e95c4e9d4174e775/src/core/packages/saved-objects/api-server/src/apis/find.ts#L94-L98) in the [dashboard search options](https://github.com/elastic/kibana/blob/main/src/platform/plugins/shared/dashboard/server/content_management/dashboard_storage.ts#L37-L47). That should be faster and less resource-intensive than manually searching the dashboards. Let me know if you want help with this.

Unfortunately, I don't think we have a similar shortcut for the `getDashboardsByField` method.

Reply:
That would be great @nickpeihl. Thank you very much for your help here.